rands.c

```diff
@@ -19,183 +20,124 @@ long _randseed = 0L;
+  \param pcg_rng is the random number generator to set.
 
- \param seed The seed.
-
- cwb - 6/27/00
 */
-void RandSeed(signed long seed) {
+void RandSeed(signed long seed, pcg32_random_t* pcg_rng) {
+//we don't need to set a random seed if RSOILWAT is used
+#ifndef RSOILWAT
 
-	if (seed == 0L) {
-		_randseed = ((long) time(NULL ));
-		if (_randseed == -1) {
-      sw_error(-1, "ERROR: RandSeed(0) called, but time() not available\n");
-		}
-		/*    _randseed %= 0xffff; */
-		_randseed *= -1;
-	} else {
-		_randseed = labs(seed) * -1;
-	}
+  if(seed == 0){
+    //seed with a random value. Uses the system time to generate
+    //a pseudo-random seed.
+    pcg32_srandom_r(pcg_rng, time(NULL), stream);
+  }
+  else {
+    //Seed with a specific value.
+    pcg32_srandom_r(pcg_rng, (int) seed, stream);
+  }
 
-#ifndef RSOILWAT
-#if RAND_FAST
-	srand(labs(_randseed));
-#endif
-#endif
+  //Increment the stream so no two generators have the same sequence.
+  stream++;
 
+#endif
 }
 
 #define BUCKETSIZE 97
 
 /*****************************************************/
 /**
-  \fn double RandUni_fast(void)
-  \brief Generate a uniform random variate.
-
-  "Fast" because it
-  utilizes the system rand() but shuffles the results
-  to make a less correlated sequence.  This code is
-  based on FUNCTION RAN0 in Press, et al., 1986,
-  Numerical Recipes, p196, Press Syndicate, NY.
-
-  Of course, just how fast is "fast" depends on the
-  implementation of the compiler.  Some older generators
-  may be quite simple, and so would be faster than
-  a more complicated algorithm.  Newer rand()'s
-  are often fast and good.
+  Random number generator using the specified rng.
 
-  cwb 18-Dec-02
-
-	\return double. A value between 0 and 1.
+  \parameters: pcg32_random_t* that specifies which stream to use.
+ 
+  \returns: a double between 0 and 1.
 */
-double RandUni_fast(void) {
-	static double y;
+double RandUni(pcg32_random_t* pcg_rng) {
 
-	#ifdef RSOILWAT
-		GetRNGstate();
-		y = unif_rand();
-		PutRNGstate();
+  double number;
 
-	#else
-		static short first_time = 1;
-		static int bucket[BUCKETSIZE];
-		static double rmax = RAND_MAX;
-		int i, j;
-
-
-		if (first_time) {
-			first_time = 0;
-			for (j = 0; j < BUCKETSIZE; j++) {
-				rand();
-			}
-			for (j = 0; j < BUCKETSIZE; j++) {
-				bucket[j] = rand();
-			}
-			y = rand() / rmax;
-		}
-		i = 1 + (int) ((double) BUCKETSIZE * y);
-		i = max(BUCKETSIZE -1, min(i,0));
-		y = bucket[i] / rmax;
-		bucket[i] = rand();
-	#endif
+#ifdef RSOILWAT
 
-	return y;
+  GetRNGstate(); 
+  number = unif_rand();
+  PutRNGstate();
+
+#else
+  // get a random unsigned int and cast it to a signed int.
+  number = (int) pcg32_random_r(pcg_rng);
+  // negative values are possible. This takes the absolute value.
+  number = fabs(number); 
+  //printf("%f\n", number/(double)RAND_MAX);
+  // divide by RAND_MAX to get a value between 0 and 1.
+  number = number / (double)RAND_MAX;

```

Comment:
`pcg32_random_r / RAND_MAX` is not guaranteed to be as expected. `RAND_MAX` is defined in `stdlib.h` and its value is implementation specific, albeit at least 32767.

`pcg32_random_r`, however, produces an integer between 0 and 2^32 (see, http://www.pcg-random.org/using-pcg-c-basic.html).
To correctly get a value between 0 and 1, you need to follow the `pcg` instructions, i.e., 
```
double d = ldexp(pcg32_random_r(&myrng), -32);
```